feat(opentelemetry-node): add @elastic/opentelemetry-instrumentation-openai

trentm opened the pull request:
- Adds the new instr to the default set.
- Adds a simple test case for it. This adds the Ollama test-service and does a minimal test of openai embeddings, because we can then depend on a small model (~45MB). We leave the complex tests to the instr-openai package itself.
- Adds a top-level examples/openai-chat.js.
- Makes a couple tweaks to EDOT to avoid log.warn's from the underlying components. (I'll add reviewer notes for those.)

trentm commented:
Running the top-level example looks like this:

```
% cd examples
% OPENAI_API_KEY=sk-... node -r @elastic/opentelemetry-node openai-chat.js | npx pino-pretty
[12:09:26.258] INFO (elastic-otel-node): start Elastic Distribution of OpenTelemetry Node.js
    preamble: true
    distroVersion: "0.5.0"
    env: {
      "os": "darwin 24.1.0",
      "arch": "arm64",
      "runtime": "Node.js v20.18.1"
    }
The sky appears blue because of Rayleigh scattering. When sunlight passes through the atmosphere, shorter blue wavelengths scatter more than longer red wavelengths, making the sky look predominantly blue to our eyes.
```

And telemetry (in mockotlpserver) looks like:

```
------ logs (1 records) ------
[2024-12-13T20:09:27.321999756Z] SEV-9 (unknown_service:node, traceId=be54f3, spanId=3a8a92): event "gen_ai.choice"
    {
      finish_reason: 'stop',
      index: 0,
      message: {}
    }
    --
    'event.name': 'gen_ai.choice',
    'gen_ai.system': 'openai'
------ trace 946ab6 (1 span) ------
       span 469499 "GET" (17.1ms, STATUS_CODE_ERROR, SPAN_KIND_CLIENT, GET http://169.254.169.254/computeMetadata/v1/instance)
------ trace 9cdd2c (1 span) ------
       span 52dbb3 "GET" (17.4ms, STATUS_CODE_ERROR, SPAN_KIND_CLIENT, GET http://metadata.google.internal./computeMetadata/v1/instance)
------ trace be54f3 (2 spans) ------
       span 3a8a92 "chat gpt-4o-mini" (1011.0ms, SPAN_KIND_CLIENT, GenAI openai, finish_reasons=stop, tokens 16in/37out)
  +8ms `- span 337d7c "POST" (998.5ms, SPAN_KIND_CLIENT, POST https://api.openai.com/v1/chat/completions -> 200)
```

(Aside: The GCP resource detector spans are still something we want to deal with.)

david-luna commented on packages/opentelemetry-node/lib/instrumentations.js:
```diff
@@ -59,6 +60,7 @@
  */
 
 /* eslint-disable prettier/prettier */
+const {OpenAIInstrumentation} = require('@elastic/opentelemetry-instrumentation-openai');
```
I guess the alphabetical order is per package scope & name? if this goes upstream we should move it down then?

trentm replied:
I guess. That's why I put it at the top. I don't have a preference though.

david-luna commented:
@trentm the openAI instrumentation has some warnings that now appear in all PRs. I'm okay withe the `any` types but the linter makes noise. Could you leverage this one to fix them or disable the linter?

trentm commented:
> @trentm the openAI instrumentation has some warnings that now appear in all PRs. I'm okay withe the `any` types but the linter makes noise. Could you leverage this one to fix them or disable the linter?

@david-luna 
There had been some lint *errors* in packages/mockotlpserver that I updated.

Do you mean the "any" lint warnings?

```

/Users/trentm/el/elastic-otel-node6/packages/instrumentation-openai/src/instrumentation.ts
   71:11  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  160:47  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  162:43  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  169:22  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  186:35  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  242:13  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  285:35  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  331:51  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  391:31  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  404:45  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  523:13  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  531:32  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  545:39  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  556:69  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  643:23  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  645:43  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  667:26  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  685:32  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  721:13  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/Users/trentm/el/elastic-otel-node6/packages/instrumentation-openai/src/utils.ts
  96:43  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
```

Do you suggest I added eslint-disable comments for all of these?

At least in opentelemetry-js-contrib.git this isn't typically done. E.g. instrumentation-bunyan has a few warnings:

```
[19:04:56 trentm@peach:~/src/opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-bunyan (git:main)]
% npm run lint

> @opentelemetry/instrumentation-bunyan@0.44.0 lint
> eslint . --ext .ts


/Users/trentm/src/opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-bunyan/src/OpenTelemetryBunyanStream.ts
  133:29  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

/Users/trentm/src/opentelemetry-js-contrib/plugins/node/opentelemetry-instrumentation-bunyan/src/instrumentation.ts
   46:18  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
   60:39  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  145:30  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
```

but those don't show up in other opentelemetry-js-contrib.git PRs, IIUC.

david-luna commented:
> but those don't show up in other opentelemetry-js-contrib.git PRs, IIUC.

Not showing up in the PR would be enough for me. Do you know how is not showing in opentelemetry-js-contrib.git?
